### PR TITLE
Cookbook ecoc random

### PR DESCRIPTION
**Pull request by OXPHOS**
- I think only one example is enough for ecoc (there're 4 examples in .py now)
- I cannot find good reference..I did find several publications but they didn't seem to be basic enough to be refed.
- Result keeps changing. So I guess random encoder is not a very good choice for the display

### CODE REVIEW

**Comment by vigsterkr** [comment]
@OXPHOS fix the random seed :)

**Comment by OXPHOS** [comment]
And maybe separate classifier/multi-class classifier in cookbook?

**Comment by OXPHOS** [comment]
@vigsterkr  I tried `math.init_random()` and it didn't work

**Comment by vigsterkr** [comment]
you mean `math.init_random(123)` ? or `math.init_random()`?

**Comment by OXPHOS** [comment]
@vigsterkr math.init_random(1)

**Comment by vigsterkr** [comment]
@OXPHOS ok but then that should be investigated why it kept giving different results, as then it's a bug. setting the seed to a fix value should make everything deterministic. can you isolate maybe the code part somehow?

**Comment by OXPHOS** [comment]
@vigsterkr I figured out why... it should be `Math:init_random(1)`..

**Comment by karlnapf** [comment]
@OXPHOS it is really important to read error messages.
"I tried `math.init_random()` and it didnt work" is not useful. Also for yourself. Try to avoid this "somehow doesnt work" mindset.

File: examples/meta/src/classifier/multiclass_ecoc_random.sg [hunk 1]
```diff
@@ -0,0 +1,39 @@
+CSVFile f_feats_train("../../data/classifier_4class_2d_linear_features_train.dat")
+CSVFile f_feats_test("../../data/classifier_4class_2d_linear_features_test.dat")
+CSVFile f_labels_train("../../data/classifier_4class_2d_linear_labels_train.dat")
+CSVFile f_labels_test("../../data/classifier_4class_2d_linear_labels_test.dat")
+Math:init_random(1)
```
Comment (karlnapf) [inline examples/meta/src/classifier/multiclass_ecoc_random.sg]: please use ß

File: doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+Multi-class Error-Correcting Output Codes
+=========================================
+
+ECOC (Error-Correcting Output Codes) is a multi-class learning strategy. ECOC trains :math:`L` binary classifers and transforms the results of the multiple classifications
+into a matrix, which is also called ECOC codebook. A decoder is applied to intepret the codebook and predict the labels of the samples.
+
+The difference between ECOC and OvR/OvO strategies (`See multi-class linear machine cookbook <http://shogun.ml/cookbook/latest/examples/classifier/multiclass_linearmachine.html>`_) 
+is that in ECOC, :math:`L` is greater than class number :math:`K`, hence the training process is error-correcting.
+
+There are multiple methods to encode or decode a codebook. In this example, we showed how to apply random encoder and hamming distance decoder to multi-class dataset.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst]: "we show"  -- present tense

File: doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+Codebook can also be encoded as dense or sparse. For dense codebook, only :math:`+1` and :math:`-1` are generated as labels for each sample in each binary classifier. While in
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst]: Codebooks

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst]: For dense codebooks

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst]: Wrong grammar. Dont start sentence with while

File: doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+into a matrix, which is also called ECOC codebook. A decoder is applied to intepret the codebook and predict the labels of the samples.
+
+The difference between ECOC and OvR/OvO strategies (`See multi-class linear machine cookbook <http://shogun.ml/cookbook/latest/examples/classifier/multiclass_linearmachine.html>`_) 
+is that in ECOC, :math:`L` is greater than class number :math:`K`, hence the training process is error-correcting.
+
+There are multiple methods to encode or decode a codebook. In this example, we showed how to apply random encoder and hamming distance decoder to multi-class dataset.
+
+Codebook can also be encoded as dense or sparse. For dense codebook, only :math:`+1` and :math:`-1` are generated as labels for each sample in each binary classifier. While in 
+sparse codebook, :math:`+1`, :math:`-1` and :math:`0` are allowed, where :math:`0` labels the samples that are not classified. 
+In this examples, we use :sgclass:`CECOCRandomDenseEncoder` as encoder. The keyword `Dense` can be replaced with `Sparse` to enable sparse encoder.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst]: There is no keyword, just mention the other class (so that it is also linked)

**Comment by karlnapf** [comment]
This is good. Put the integration test data in, correct minor comments, squash and we can merge

**Comment by karlnapf** [comment]
http://www.jmlr.org/papers/volume11/escalera10a/escalera10a.pdf

**Comment by OXPHOS** [comment]
@karlnapf I don't think `init_random(0)` works because the implementation of `math.init_random()` is:
```
			if (initseed==0)
				seed = CRandom::generate_seed();
			else
				seed=initseed;
```
and I did several tests and the results indeed kept changing with `init_random(0)`.

**Comment by vigsterkr** [comment]
@OXPHOS WAIT A FUCKING SECOND.
so did you try `math.init_random(1)` or `math.init_random(0)`.

in https://github.com/shogun-toolbox/shogun/pull/3280#issuecomment-225344176 you said `math.init_random(1)` hence my surprise

File: examples/meta/src/classifier/multiclass_ecoc_random.sg (hunk 1 above)
Comment (karlnapf) [inline examples/meta/src/classifier/multiclass_ecoc_random.sg]: EDIT:
the 0 seed in Shogun causes a random seed to be generated. I didnt know that
So you have to use 1.

Also, can you send a mini PR that adds a comment in the CMath interface documentaion?

**Comment by karlnapf** [comment]
All this confusion is actually due to me, as I didnt know about our seed=0 behaviour.
Solution for all of this: use seed 1 :)

**Comment by OXPHOS** [comment]
@vigsterkr I tried both `math.init_random(1)` and `math.init_random(0)` in the beginning - The grammar of these were actually wrong - I was confused because it DIDN'T raise any error - I didn't make properly I guess.
The correct way is `Math:init_random(1)`.

@karlnapf I had the test data pulled and I'll update once that is merged.

**Comment by sorig** [comment]
@OXPHOS the meta language does not know what methods the classes have (and whether they're static methods or not). It therefore can't raise errors on mistakes like this. 

```
math.init_random(1)
```

Is grammatically correct to the meta language, but not semantically correct as the class is called `Math` and `init_random` is a static method. This must be checked downstream once the examples are run in the target languages. I.e. it's not enough just to check if the programs  parse - it's important to actually translate them, run them, and then inspect any error messages.

File: doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst [hunk 5]
```diff
@@ -0,0 +1,54 @@
+=========================================
+Multi-class Error-Correcting Output Codes
+=========================================
+
+ECOC (Error-Correcting Output Codes) is a multi-class learning strategy. ECOC trains :math:`L` binary classifers and transforms the results of the multiple classifications
+into a matrix, which is also called ECOC codebook. A decoder is applied to intepret the codebook and predict the labels of the samples.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst]: remove "also"

Comment (karlnapf) [inline doc/cookbook/source/examples/classifier/multiclass_ecoc_random.rst]: "codebook, and to predict"

**Comment by karlnapf** [comment]
All green, merging. Please send a mini update for the minor comments

**Comment by OXPHOS** [comment]
@sorig Thanks for the advice!

**Comment by OXPHOS** [comment]
@karlnapf  actually I was still wondering..

- There're 4 ecoc examples in undocumented now. Three of them just used different encoder and might be able to merged into current example (i.e. the .py can be removed now). The other one is a comparison between the efficiency of different ecoc encoders and decoders. Don't know whether it should be in the cookbook..

- As there are tons of classifier methods, do you think we can separate `classifier` and `multi-class classifier` group?

**Comment by karlnapf** [comment]
Leave the comparison example in, remove the others.
But for the others, please extend the existing cookbook

Yeah we can separate multiclass from classifer.
Binary classifier
Multiclass classifer
Feel free to send a mini PR with adjustment for that. Thanks!

**Pull request merged**